### PR TITLE
Added RunRuleAction for AutomationComponent

### PR DESCRIPTION
**Pull request by BenediktNiehues**
Implementation for issue:  "create a RunRuleAction #2964"
Signed-off-by: Benedikt Niehues <benedikt.niehues@itemis.de>

### CODE REVIEW

File: bundles/automation/org.eclipse.smarthome.automation.module.core/src/main/java/org/eclipse/smarthome/automation/module/core/handler/RunRuleActionHandler.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+import org.eclipse.smarthome.automation.handler.ActionHandler;
+import org.eclipse.smarthome.automation.handler.BaseModuleHandler;
+import org.eclipse.smarthome.config.core.Configuration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is a handler for RunRuleAction module type.
+ * It runs the rules which's UIDs are passed by the 'ruleUIDs' property.
+ * !!! If a rule's status is not IDLE that rule can not run !!!
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.module.core/src/main/java/org/eclipse/smarthome/automation/module/core/handler/RunRuleActionHandler.java]: I don't think we need `!!!` here in front of the sentence and after it.
A single `!` at the end should be enough. Or use `must` / `must not`.

File: bundles/automation/org.eclipse.smarthome.automation.module.core/src/main/java/org/eclipse/smarthome/automation/module/core/handler/RunRuleActionHandler.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+    /**
+     * the UIDs of the rules to be executed.
+     */
+    private final List<String> ruleUIDs;
+    
+    /**
+     * boolean to express if the conditions should be considered, defaults to true;
+     */
+    private boolean considerConditions=true;
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.module.core/src/main/java/org/eclipse/smarthome/automation/module/core/handler/RunRuleActionHandler.java]: I assume there is something wrong with your formatter. It should add a space in front of / after `=`

File: bundles/automation/org.eclipse.smarthome.automation.module.core/src/main/java/org/eclipse/smarthome/automation/module/core/handler/RunRuleActionHandler.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+    @SuppressWarnings("unchecked")
+    public RunRuleActionHandler(final Action module, final RuleRegistry ruleRegistry) {
+        super(module);
+        final Configuration config = module.getConfiguration();
+        if (config == null) {
+            throw new IllegalArgumentException("'Configuration' can not be null.");
+        }
+
+		ruleUIDs = (List<String>) config.get(RULE_UIDS_KEY);
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.module.core/src/main/java/org/eclipse/smarthome/automation/module/core/handler/RunRuleActionHandler.java]: Tabs? Your formatter is not working.

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java [hunk 4]
```diff
@@ -844,36 +844,46 @@ protected void runRule(RuntimeRule rule, RuleEngineCallbackImpl.TriggerData td)
         }
     }
 
+    protected void runNow(String ruleUID, boolean considerConditions){
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java]: code formatter missing? At least there should be a space before the curly bracket and you have a lot of tabs in here...

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java [hunk 5]
```diff
@@ -844,36 +844,46 @@ protected void runRule(RuntimeRule rule, RuleEngineCallbackImpl.TriggerData td)
...
+    	 RuntimeRule rule = getRuntimeRule(ruleUID);
+         if (rule == null) {
+             logger.warn("Fail to execute rule '{}': {}", ruleUID, "Invalid Rule UID.");
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java]: I see that it is not from your PR, but could you please fix these?

typo: Failed

Don't make "Invalid Rule UID." a parameter, but add it to the logging string itself.

File: bundles/automation/org.eclipse.smarthome.automation.module.core.test/src/test/groovy/org/eclipse/smarthome/automation/module/core/internal/RunRuleModuleTest.groovy [hunk 6]
```diff
@@ -0,0 +1,161 @@
+/**
+ * Copyright (c) 1997, 2015 by ProSyst Software GmbH and others.
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.module.core.test/src/test/groovy/org/eclipse/smarthome/automation/module/core/internal/RunRuleModuleTest.groovy]: you can put 2017 and Deutsche Telekom AG here

File: bundles/automation/org.eclipse.smarthome.automation.module.core.test/src/test/groovy/org/eclipse/smarthome/automation/module/core/internal/RunRuleModuleTest.groovy [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    @Before
+    void before() {
+
+        def itemProvider = [
+            getAll: {
+                [
+                    new SwitchItem("switch1"),
+                    new SwitchItem("switch2"),
+                    new SwitchItem("switch3"),
+					new SwitchItem("ruleTrigger")
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.module.core.test/src/test/groovy/org/eclipse/smarthome/automation/module/core/internal/RunRuleModuleTest.groovy]: code formatter can't be bad here either...

File: bundles/automation/org.eclipse.smarthome.automation.module.core/ESH-INF/automation/moduletypes/RunRuleAction.json [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+			"description": "runs/executes a rule or a group of rules specified by their UIDs.",
+			"configDescriptions": [
+				{
+					"name": "ruleUIDs",
+					"type": "TEXT",
+					"label": "Rules",
+					"description": "Specifies the target rule(s) that should be executed.",
+					"required": true,
+					"multiple": true,
+					"context":"rule"
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.module.core/ESH-INF/automation/moduletypes/RunRuleAction.json]: add a space after the colon

File: bundles/automation/org.eclipse.smarthome.automation.module.core/ESH-INF/automation/moduletypes/RunRuleAction.json [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+				},
+				{
+					"name": "considerConditions",
+					"type": "BOOLEAN",
+					"label": "consider Conditions",
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.module.core/ESH-INF/automation/moduletypes/RunRuleAction.json]: Better "Consider Conditions"

File: bundles/automation/org.eclipse.smarthome.automation.module.core/ESH-INF/automation/moduletypes/RunRuleAction.json [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+					"description": "Specifies whether the conditions of the target rule(s) to be executed should be considered or not.",
+					"required": true,
+					"default": true,
+					"options": [
+						{
+							"label": "Enable",
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.module.core/ESH-INF/automation/moduletypes/RunRuleAction.json]: wouldn't the better selection be "Consider conditions" -> Yes/No and not Enable/Disable?

File: bundles/automation/org.eclipse.smarthome.automation.module.core/src/main/java/org/eclipse/smarthome/automation/module/core/handler/RunRuleActionHandler.java [hunk 11]
```diff
@@ -0,0 +1,105 @@
+/**
+ * Copyright (c) 1997, 2016 by ProSyst Software GmbH and others.
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.module.core/src/main/java/org/eclipse/smarthome/automation/module/core/handler/RunRuleActionHandler.java]: you can put 2017 and Deutsche Telekom AG here

File: bundles/automation/org.eclipse.smarthome.automation.module.core/src/main/java/org/eclipse/smarthome/automation/module/core/handler/RunRuleActionHandler.java [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+    /**
+     * the key for the 'rulesUIDs' property of the {@link Action}.
+     */
+    private final static String RULE_UIDS_KEY = "ruleUIDs";
+    private final static String CONSIDER_CONDITIONS_KEY = "considerConditions";
+
+    /**
+     * The logger
+     */
+    private final static Logger logger = LoggerFactory.getLogger(RunRuleActionHandler.class);
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.module.core/src/main/java/org/eclipse/smarthome/automation/module/core/handler/RunRuleActionHandler.java]: do you need it static?

File: bundles/automation/org.eclipse.smarthome.automation.module.core/src/main/java/org/eclipse/smarthome/automation/module/core/handler/RunRuleActionHandler.java [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+    public RunRuleActionHandler(final Action module, final RuleRegistry ruleRegistry) {
+        super(module);
+        final Configuration config = module.getConfiguration();
+        if (config == null) {
+            throw new IllegalArgumentException("'Configuration' can not be null.");
+        }
+
+        ruleUIDs = (List<String>) config.get(RULE_UIDS_KEY);
+        if (ruleUIDs == null) {
+            throw new IllegalArgumentException("'ruleUIDs' property can not be null.");
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.module.core/src/main/java/org/eclipse/smarthome/automation/module/core/handler/RunRuleActionHandler.java]: it seems it "can", so better say it "must not" ;-)

File: bundles/automation/org.eclipse.smarthome.automation.module.core/src/main/java/org/eclipse/smarthome/automation/module/core/handler/RunRuleActionHandler.java [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    @Override
+    public Map<String, Object> execute(Map<String, Object> context) {
+        // execute each rule after the other; at the moment synchronously
+        for (String uid : ruleUIDs) {
+            if (ruleRegistry != null) {
+                ruleRegistry.runNow(uid, considerConditions);
+            } else {
+                logger.warn("Action is not applyed to {} because RuleRegistry is not available.", uid);
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.module.core/src/main/java/org/eclipse/smarthome/automation/module/core/handler/RunRuleActionHandler.java]: typo: applied

File: bundles/automation/org.eclipse.smarthome.automation.module.core/src/main/java/org/eclipse/smarthome/automation/module/core/factory/CoreModuleHandlerFactory.java [hunk 15]
```diff
@@ -182,6 +183,8 @@ protected synchronized ModuleHandler internalCreate(final Module module, final S
                 return postCommandActionHandler;
             } else if (RuleEnablementActionHandler.UID.equals(moduleTypeUID)) {
                 return new RuleEnablementActionHandler((Action) module, ruleRegistry);
+            } else if (RunRuleActionHandler.UID.equals(moduleTypeUID)){
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.module.core/src/main/java/org/eclipse/smarthome/automation/module/core/factory/CoreModuleHandlerFactory.java]: If you could add a space before the opening bracket, I would be fine to merge the PR :-)

**Pull request merged**